lib/maths/CPeriodicityHypothesisTests.cc

```diff
@@ -1376,167 +1587,224 @@ bool CPeriodicityHypothesisTests::testPeriod(const TTimeTimePr2Vec& windows,
     TTimeTimePr2Vec window{{0, windowLength}};
+
+    // Get the values minus the trend for the null hypothesis.
     TFloatMeanAccumulatorVec values(buckets.begin(), buckets.end());
     this->conditionOnHypothesis(stats, values);
 
+    // If necessary project the set of all buckets onto the set which
+    // intersect the windows.
     if (windowLength < length(buckets, m_BucketLength)) {
         LOG_TRACE(<< "Projecting onto " << core::CContainerPrinter::print(windows));
         TFloatMeanAccumulatorVec projection;
         project(values, windows, m_BucketLength, projection);
         values = std::move(projection);
     }
 
-    double B{static_cast<double>(std::count_if(
+    // Compute the number of non-empty buckets. This needs to account for
+    // projecting onto windows.
+    stats.s_B = static_cast<double>(std::count_if(
         values.begin(), values.end(), [](const TFloatMeanAccumulator& value) {
             return CBasicStatistics::count(value) > 0.0;
-        }))};
-    double df0{B - stats.s_DF0};
+        }));
 
     // We need fewer degrees of freedom in the null hypothesis trend model
     // we're fitting than non-empty buckets.
-    if (df0 <= 0.0) {
+    if (stats.s_B <= stats.s_DF0) {
         return false;
     }
 
+    // Fit the periodic trend and compute the degrees of freedom given the
+    // alternative hypothesis.
     TMeanVarAccumulatorVec trend(period);
     periodicTrend(values, window, m_BucketLength, trend);
-
     double b{static_cast<double>(std::count_if(
         trend.begin(), trend.end(), [](const TMeanVarAccumulator& value) {
             return CBasicStatistics::count(value) > 0.0;
         }))};
-    double df1{B - b};
+    double df1{stats.s_B - b};
     LOG_TRACE(<< "  populated = " << b);
 
-    // We need fewer degrees of freedom in the trend model we're fitting
-    // than non-empty buckets.
+    // We need fewer points in the trend model we're fitting than non-empty
+    // buckets.
     if (df1 <= 0.0) {
         return false;
     }
 
+    // Compute the residual variance in the alternative hypothesis *without*
+    // removing outliers.
     double scale{1.0 / stats.s_M};
     LOG_TRACE(<< "  scale = " << scale);
-
     double v{residualVariance<double>(trend, scale)};
     v = varianceAtPercentile(v, df1, 50.0 + CONFIDENCE_INTERVAL / 2.0);
+
+    // Fit the periodic trend re-weighting outliers.
     reweightOutliers(trend, window, m_BucketLength, values);
     trend.assign(period, TMeanVarAccumulator{});
     periodicTrend(values, window, m_BucketLength, trend);
+    double v1{residualVariance<double>(trend, scale)};
+
+    double R;
+    double meanRepeats;
+    double pVariance;
+    return this->testVariance(window, values, period_, df1, v1, stats, R,
+                              meanRepeats, pVariance) ||
+           this->testAmplitude(window, values, period_, b, v, R, meanRepeats,
+                               pVariance, stats);
```

Comment:
I factored out `testVariance` and `testAmplitude` from this function (for reuse in `CPeriodicityHypothesisTests::testPeriodWithScaling`). These are what is being compared to the following code.